### PR TITLE
Actor LTS

### PR DESCRIPTION
**Pull request by krenzland**
This is a refactoring (or rather a rewrite) of our Time Cluster machinery. It includes:

- Step-based criteria instead of time-based scheduling: Avoids floating point comparisons
- New code structure for time cluster
- A refactoring of the communication thread

I've checked a lot of scenarios but I would appreciate if you would also check our applications. It's quite hard to thoroughly test every feature.
@ravil-mobile Could you please check whether the GPU version still works? I've ran some tests but it would be great if you'd double check.

PS: It's a slightly to get this to compile on SuperMUC-NG. Either use the GCC compiler or load module "intel-oneapi/2021.4". Older versions of the Intel compiler crash during compilation. With the new Intel compiler the compilation works but you have use the old versions of the modules to run it. Otherwise it crashes.

### CODE REVIEW

**Comment by Thomas-Ulrich** [comment]
A very basic question, how would you define a layer?
What is the difference between a ghost, copy, and interior layer?

**Comment by krenzland** [comment]
Basically, for serial code only interior layers are relevant.
For MPI, we need all three:

- Interior cells are all cells that do not have any neighbor that is on another MPI rank
-  Copy layer are all cells that have neighbors on another rank
- Ghost layers basically represent all cells that directly neigbor cells of the copy layer.

Only interior/copy layer cells are computed on the local rank; ghost cells are computed on other ranks and are just copied.
Copy cells may be duplicate; if a cell has neighbors on two different ranks.

Interior/copy are only separated for computational reasons. We compute copy cells first and send the data directly to neighboring ranks. In this way, we can hide the communication behind the computation.

The situation with dynamic rupture is a bit more complicated but follows a similar strategy.

**Comment by Thomas-Ulrich** [comment]
Ok, thanks, that's very clear now!

**Review by sebwolf-de — COMMENTED** [review]
Just some small remarks.
A setup with poroelasticity and a point source works @Thomas-Ulrich, if one of your scenarios works, we can merge, IMHO.

File: src/Solver/time_stepping/AbstractTimeCluster.cpp [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    }
+    case ActorState::Synced:
+      if (ct.stepsSinceLastSync == 0) {
+        return ActorAction::RestartAfterSync;
+      }
+      break;
+    default:
+      throw;
+  }
+  return ActorAction::Nothing;
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/Solver/time_stepping/AbstractTimeCluster.cpp:44]: The return actually makes sense. The default case is in principle not needed - there are only three cluster states. The nothing action is a default action that is taken when no valid action is found. E.g. when ActorState::Predicted and mayCorrect() returns false.

File: src/Solver/time_stepping/AbstractTimeCluster.cpp [hunk 2]
```diff
@@ -92,8 +93,7 @@ void AbstractTimeCluster::unsafePerformAction(ActorAction action) {
     case ActorAction::Sync:
       assert(state == ActorState::Corrected);
       logDebug(MPI::mpi.rank()) << "synced at " << syncTime
-                                << ", corrTIme =" << ct.correctionTime
-                                << " stepsSinceLastSync " << ct.stepsSinceLastSync
+                                << ", corrTime =" << ct.correctionTime
                                 << " stepsUntilLastSync " << ct.stepsSinceLastSync
```
Comment (sebwolf-de) [inline src/Solver/time_stepping/AbstractTimeCluster.cpp:97]: The string says `stepsUntilLastSync`, but the outputted variable says `stepsSinceLastSync`.

File: src/Solver/time_stepping/AbstractTimeCluster.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply src/Solver/time_stepping/AbstractTimeCluster.cpp]: Okay, now it makes sense :+1:

**Review by Thomas-Ulrich — COMMENTED** [review]
obviously, I m far from understanding everything (anything?), but here are some comments.

File: src/tests/Solver/time_stepping/AbstractTimeCluster.t.h [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  bool isFinished = false;
+  auto iteration = 0;
+  while(!isFinished) {
+    isFinished = true;
+
+    ALLOW_CALL(cluster1, handleAdvancedCorrectionTimeMessage(ANY(NeighborCluster)));
+    ALLOW_CALL(cluster2, handleAdvancedCorrectionTimeMessage(ANY(NeighborCluster)));
+    ALLOW_CALL(cluster1, handleAdvancedPredictionTimeMessage(ANY(NeighborCluster)));
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/tests/Solver/time_stepping/AbstractTimeCluster.t.h:81]: Different clusters

File: src/tests/Solver/time_stepping/AbstractTimeCluster.t.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    cluster->updateSyncTime(endTime);
+    cluster->reset();
+  }
+
+  ALLOW_CALL(cluster1, handleAdvancedCorrectionTimeMessage(ANY(NeighborCluster)));
+  ALLOW_CALL(cluster2, handleAdvancedCorrectionTimeMessage(ANY(NeighborCluster)));
+  ALLOW_CALL(cluster1, handleAdvancedPredictionTimeMessage(ANY(NeighborCluster)));
+  ALLOW_CALL(cluster2, handleAdvancedPredictionTimeMessage(ANY(NeighborCluster)));
+
+  // First, move from synced -> corrected -> predicted
```
[the remark this answers is not in the document]
Comment (krenzland) [reply src/tests/Solver/time_stepping/AbstractTimeCluster.t.h:139]: Prediction = compute local integral
Correction = ocmpute neighbor integral

The lines here do not imply an ordering, they just allow that the function is called

[the remark this answers is not in the document]
Comment (Thomas-Ulrich) [reply src/tests/Solver/time_stepping/AbstractTimeCluster.t.h:139]: ok so making a correction on the local cell is equivalent to making a prediction on the neighbor cells?
(these 2 represent only 1 action, not 2?)

File: src/Initializer/tree/Lut.cpp [hunk 5]
```diff
@@ -138,22 +139,53 @@ void seissol::initializers::Lut::createLuts(  LTSTree*        ltsTree,
       maskedLut.createLut(mask, m_ltsTree, ltsToMesh, numberOfMeshIds );
     }
   }
-  
-  unsigned* clusters = new unsigned[m_ltsTree->numChildren()+1];
+
+  struct LayerCounter {
+      unsigned numGhost;
```
Comment (Thomas-Ulrich) [inline src/Initializer/tree/Lut.cpp:144]: I think it is should be offsetGhost not numGhost.

File: src/Solver/time_stepping/AbstractTimeCluster.cpp [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+AbstractTimeCluster::AbstractTimeCluster(double maxTimeStepSize, long timeStepRate)
+    : timeStepRate(timeStepRate), numberOfTimeSteps(0),
+    lastStateChange(std::chrono::steady_clock::now()) {
+  ct.maxTimeStepSize = maxTimeStepSize;
+  ct.timeStepRate = timeStepRate;
+}
+
+ActorAction AbstractTimeCluster::getNextLegalAction() {
```
[the remark this answers is not in the document]
Comment (Thomas-Ulrich) [reply src/Solver/time_stepping/AbstractTimeCluster.cpp:19]: ok I see.

**Comment by Thomas-Ulrich** [comment]
Yes, I would also merge.

**Review by Thomas-Ulrich — APPROVED** [review]

**Pull request merged**